public/app/plugins/datasource/prometheus/components/monaco-query-field/monaco-completion-provider/completions.ts

```diff
@@ -0,0 +1,140 @@
+    label: f.label,
+    insertText: f.insertText ?? '', // i don't know what to do when this is nullish. it should not be.
+  }));
+}
+
+function getAllDurationsCompletions(): Completion[] {
+  // FIXME: get a better list
+  return ['5m', '1m', '30s', '15s'].map((text) => ({
+    label: text,
+    insertText: text,
+  }));
+}
+
+async function getAllHistoryCompletions(dataProvider: DataProvider): Promise<Completion[]> {
+  // function getAllHistoryCompletions(queryHistory: PromHistoryItem[]): Completion[] {
+  // NOTE: the typescript types are wrong. historyItem.query.expr can be undefined
+  const allHistory = await dataProvider.getHistory();
+  // FIXME: find a better history-limit
+  return allHistory.slice(0, 10).map((expr) => ({
+    label: expr,
+    insertText: expr,
+  }));
+}
+
+function makeSelector(metricName: string | undefined, labels: Label[]): string {
+  const allLabels = [...labels];
+
+  // we transform the metricName to a label, if it exists
+  if (metricName !== undefined) {
+    allLabels.push({ name: '__name__', value: metricName });
+  }
+
+  const allLabelTexts = allLabels.map((label) => `${label.name}="${label.value}"`);
+
+  return `{${allLabelTexts.join(',')}}`;
+}
+
+async function getLabelNamesForCompletions(
+  metric: string | undefined,
+  suffix: string,
+  triggerOnInsert: boolean,
+  otherLabels: Label[],
+  dataProvider: DataProvider
+): Promise<Completion[]> {
+  const selector = makeSelector(metric, otherLabels);
+  const data = await dataProvider.getSeries(selector);
+  const possibleLabelNames = Object.keys(data); // all names from prometheus
+  const usedLabelNames = new Set(otherLabels.map((l) => l.name)); // names used in the query
+  const labelNames = possibleLabelNames.filter((l) => !usedLabelNames.has(l));
+  return labelNames.map((text) => ({
+    label: text,
+    insertText: `${text}${suffix}`,
+    triggerOnInsert,
+  }));
+}
+
+async function getLabelNamesForSelectorCompletions(
+  metric: string | undefined,
+  otherLabels: Label[],
+  dataProvider: DataProvider
+): Promise<Completion[]> {
+  return getLabelNamesForCompletions(metric, '=', true, otherLabels, dataProvider);
+}
+async function getLabelNamesForByCompletions(
+  metric: string | undefined,
+  otherLabels: Label[],
+  dataProvider: DataProvider
+): Promise<Completion[]> {
+  return getLabelNamesForCompletions(metric, '', false, otherLabels, dataProvider);
+}
+
+async function getLabelValuesForMetricCompletions(
+  metric: string | undefined,
+  labelName: string,
+  otherLabels: Label[],
+  dataProvider: DataProvider
+): Promise<Completion[]> {
+  const selector = makeSelector(metric, otherLabels);
+  const data = await dataProvider.getSeries(selector);
+  const values = data[labelName] ?? [];
+  return values.map((text) => ({
+    label: text,
+    insertText: `"${text}"`, // FIXME: escaping strange characters?
+  }));
+}
+
+export async function getCompletions(intent: Intent, dataProvider: DataProvider): Promise<Completion[]> {
```

Comment:
Something probably worth adding to a to-do list. Label values list contains duplicated values:

<img width="750" alt="Screen Shot 2021-08-30 at 10 45 23" src="https://user-images.githubusercontent.com/745532/131312648-0fb11b70-25c6-4141-9a2f-3bad801defd1.png">

Reply:
Can't reproduce it now though 🤔 might some sort of edge case.

Reply:
interesting. i added it to the todo-list. thanks.